### ISSUE TITLE
Bad, Sad Error when term msg is not given in "test:assertTrue(value, msg = "AssertTrue failed")" 

### ISSUE THREAD

**Issue by 1akshitha** [ballerina-platform/ballerina-lang#13578]
**Description:**
Bad sad error when term msg is not given in "test:assertTrue(value, msg = "AssertTrue failed")" 

**Steps to reproduce:**
Remove term msg.
@test:Config
function testAssertTrue() {
    boolean value = true;
    test:assertTrue(value, "AssertTrue failed");
}

**Affected Versions:**

Ballerina 0.990.2

**Error log:**

[2019-02-12 10:34:31,722] ERROR {org.ballerinalang.launcher.Main} - null 
java.lang.NullPointerException
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.validateBinaryExpr(CodeAnalyzer.java:1537)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.visit(CodeAnalyzer.java:1528)
	at org.wso2.ballerinalang.compiler.tree.expressions.BLangBinaryExpr.accept(BLangBinaryExpr.java:60)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.analyzeExpr(CodeAnalyzer.java:1849)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.analyzeExprs(CodeAnalyzer.java:1889)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.visit(CodeAnalyzer.java:1575)
	at org.wso2.ballerinalang.compiler.tree.expressions.BLangBracedOrTupleExpr.accept(BLangBracedOrTupleExpr.java:48)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.analyzeExpr(CodeAnalyzer.java:1849)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.analyzeExprs(CodeAnalyzer.java:1889)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.visit(CodeAnalyzer.java:1341)
	at org.wso2.ballerinalang.compiler.tree.expressions.BLangInvocation.accept(BLangInvocation.java:112)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.analyzeExpr(CodeAnalyzer.java:1849)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.visit(CodeAnalyzer.java:1072)
	at org.wso2.ballerinalang.compiler.tree.statements.BLangExpressionStmt.accept(BLangExpressionStmt.java:47)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.analyzeNode(CodeAnalyzer.java:276)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.lambda$visit$3(CodeAnalyzer.java:467)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.visit(CodeAnalyzer.java:467)
	at org.wso2.ballerinalang.compiler.tree.statements.BLangBlockStmt.accept(BLangBlockStmt.java:54)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.analyzeNode(CodeAnalyzer.java:276)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.visit(CodeAnalyzer.java:487)
	at org.wso2.ballerinalang.compiler.tree.statements.BLangIf.accept(BLangIf.java:66)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.analyzeNode(CodeAnalyzer.java:276)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.lambda$visit$3(CodeAnalyzer.java:467)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.visit(CodeAnalyzer.java:467)
	at org.wso2.ballerinalang.compiler.tree.statements.BLangBlockStmt.accept(BLangBlockStmt.java:54)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.analyzeNode(CodeAnalyzer.java:276)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.visitFunction(CodeAnalyzer.java:349)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.visit(CodeAnalyzer.java:327)
	at org.wso2.ballerinalang.compiler.tree.BLangFunction.accept(BLangFunction.java:68)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.analyzeNode(CodeAnalyzer.java:276)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.lambda$analyzeTopLevelNodes$1(CodeAnalyzer.java:265)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.analyzeTopLevelNodes(CodeAnalyzer.java:265)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.visit(CodeAnalyzer.java:260)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.lambda$visit$0(CodeAnalyzer.java:261)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.visit(CodeAnalyzer.java:261)
	at org.wso2.ballerinalang.compiler.tree.BLangPackage.accept(BLangPackage.java:159)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.analyze(CodeAnalyzer.java:249)
	at org.wso2.ballerinalang.compiler.CompilerDriver.codeAnalyze(CompilerDriver.java:203)
	at org.wso2.ballerinalang.compiler.CompilerDriver.compile(CompilerDriver.java:157)
	at org.wso2.ballerinalang.compiler.CompilerDriver.compilePackageSymbol(CompilerDriver.java:144)
	at org.wso2.ballerinalang.compiler.CompilerDriver.compilePackage(CompilerDriver.java:106)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
	at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:175)
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1382)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418)
	at org.wso2.ballerinalang.compiler.Compiler.compilePackages(Compiler.java:161)
	at org.wso2.ballerinalang.compiler.Compiler.compilePackage(Compiler.java:179)
	at org.wso2.ballerinalang.compiler.Compiler.compile(Compiler.java:88)
	at org.wso2.ballerinalang.compiler.Compiler.build(Compiler.java:100)
	at org.ballerinalang.packerina.BuilderUtils.compileWithTestsAndWrite(BuilderUtils.java:74)
	at org.ballerinalang.packerina.cmd.BuildCommand.execute(BuildCommand.java:176)
	at java.util.Optional.ifPresent(Optional.java:159)
	at org.ballerinalang.launcher.Main.main(Main.java:67)

**Comment by MaryamZi** [comment]
Would it be possible to share the complete source?

Tried the following with both 0.990.2 and 0.990.3 (both `run` and `test`) but got the compilation failure as expected:
```ballerina
// err_test.bal
import ballerina/test;

@test:Config
function testAssertTrue() {
    boolean value = true;
    test:assertTrue(value, "AssertTrue failed");
}
```

```cmd
$ ballerina test err_test.bal
Compiling tests
    err_test.bal
error:.::err_test.bal:7:5: defaultable parameter expected. defaultable parameters should be passed as key-value pairs
```

Tested with a multi-module project too.

**Comment by 1akshitha** [comment]
https://github.com/1akshitha/module-google-natural-language/blob/master/google_natural_language_v1/tests/test.bal
 
Line 29

**Comment by MaryamZi** [comment]
Thank you @1akshitha. 

This seems to be happening when binary expressions are used within the type guard, where there is a compilation error. 

Simple code to reproduce:
```bal
public function main() {
    any a = 1;
    if (a is int) {
        // try to call the function with too few arguments
        takesTwo(a == 0);
    }
}

function takesTwo(boolean b, string s) {
    // do nothing
}
```

```cmd
$ ballerina run err_test.bal
error: .::err_test.bal:5:9: not enough arguments in call to 'takesTwo()'
ballerina: Oh no, something really went wrong. Bad. Sad.
...
```

**Comment by MaryamZi** [comment]
This is fixed in 0.992.0-m8.

The following errors are printed.
```cmd
$ ballerina run err_test.bal
error: .::err_test.bal:5:9: not enough arguments in call to 'takesTwo()'
error: .::err_test.bal:5:9: missing required parameter 's' in call to 'takesTwo'()
```
